frontend/app/controllers/spree/coupon_codes_controller.rb

```diff
@@ -17,9 +19,12 @@ def create
             if handler.successful?
               flash[:success] = handler.success
               redirect_to cart_path
-            else
-              flash.now[:error] = handler.error
+            elsif handler.errors.count == 1
```

Comment:
I have mixed feelings here but I probably prefer to use a generic solution (a generic message for failed coupon code application and always relying on the views for explaining details). I get why you think this is the best approach but I think we can make this change safely, it's not a big deal for backward compatibility to show a custom message here and we can save some `if`s in the view as well.